File: backend/app/schemas/usage.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class DailyStat(BaseModel):
    date: str
    calls: int
    tokens: int
    cost_cents: int = Field(..., alias="costCents")

    model_config = {"populate_by_name": True}


class UsageStatsResponse(BaseModel):
    today_calls: int = Field(..., alias="todayCalls")
    today_tokens: int = Field(..., alias="todayTokens")
    today_cost_cents: int = Field(..., alias="todayCostCents")
    active_keys: int = Field(..., alias="activeKeys")
    daily: list[DailyStat]

    model_config = {"populate_by_name": True}


class UsageRecordResponse(BaseModel):
    id: int
    model: str
    request_tokens: int = Field(..., alias="requestTokens")
    response_tokens: int = Field(..., alias="responseTokens")
    total_tokens: int = Field(..., alias="totalTokens")
    cost_cents: int = Field(..., alias="costCents")
    created_at: datetime = Field(..., alias="createdAt")

    model_config = {"populate_by_name": True}


class UsageHistoryResponse(BaseModel):
    records: list[UsageRecordResponse]
    total: int
    page: int
    page_size: int = Field(..., alias="pageSize")

    model_config = {"populate_by_name": True}

File: frontend/lib/api/usage.ts
import { useQuery } from "@tanstack/react-query";
import { apiClient, ApiClientError } from "./client";

// --- Types ---

export interface DailyStat {
  date: string;
  calls: number;
  tokens: number;
  costCents: number;
}

export interface UsageStats {
  todayCalls: number;
  todayTokens: number;
  todayCostCents: number;
  activeKeys: number;
  daily: DailyStat[];
}

export interface UsageRecord {
  id: number;
  model: string;
  requestTokens: number;
  responseTokens: number;
  totalTokens: number;
  costCents: number;
  createdAt: string;
}

export interface UsageHistory {
  records: UsageRecord[];
  total: number;
  page: number;
  pageSize: number;
}

// --- Hooks ---

export function useUsageStats(days: number = 30) {
  return useQuery<UsageStats, ApiClientError>({
    queryKey: ["usage", "stats", days],
    queryFn: () =>
      apiClient<UsageStats>(`/usage/stats?days=${days}`),
    refetchInterval: 60_000, // Auto-refresh every 60 seconds
  });
}

export function useUsageHistory(page: number = 1, pageSize: number = 20) {
  return useQuery<UsageHistory, ApiClientError>({
    queryKey: ["usage", "history", page, pageSize],
    queryFn: () =>
      apiClient<UsageHistory>(
        `/usage/history?page=${page}&pageSize=${pageSize}`
      ),
  });
}
